pinot-core/src/main/java/com/linkedin/pinot/core/realtime/impl/dictionary/BaseOffHeapMutableDictionary.java

```diff
@@ -0,0 +1,476 @@
+    if (numBuffers == 0) {
+      return expand(_sizeOfFirstBuf);
+    }
+    final int lastCapacity = iBufList.get(numBuffers-1).capacity();
+    int expansionMultiple = EXPANSION_MULTIPLES[EXPANSION_MULTIPLES.length-1];
+    if (numBuffers < EXPANSION_MULTIPLES.length) {
+      expansionMultiple = EXPANSION_MULTIPLES[numBuffers];
+    }
+    int newSize = validateSize(lastCapacity, expansionMultiple);
+    return expand(newSize);
+  }
+
+  private int validateSize(int lastCapacity, int expansionMultiple) {
+    long newSize = (long) lastCapacity * expansionMultiple;
+    while (newSize >= Integer.MAX_VALUE) {
+      newSize = newSize/2;
+    }
+    return nearestPrime((int) newSize);
+  }
+
+  protected int nearestPowerOf2(int num) {
+    if ((num & (num -1)) == 0) {
+      return num;
+    }
+    int power = Integer.SIZE - Integer.numberOfLeadingZeros(num);
+    Preconditions.checkState(power < Integer.SIZE-1);
+    return 1 << power;
+  }
+
+  protected int getDictId(@Nonnull Object rawValue) {
+    final long hashVal = Math.abs(rawValue.hashCode());
+    final ValueToDictId valueToDictId = _valueToDict.get();
+    final List<IntBuffer> iBufList = valueToDictId.getIBufList();
+    for (IntBuffer iBuf : iBufList) {
+      final int modulo = iBuf.capacity()/NUM_COLUMNS;
+      final int offsetInBuf = (int)(hashVal % modulo)  * NUM_COLUMNS;
+      for (int i = offsetInBuf; i < offsetInBuf + NUM_COLUMNS; i++) {
+        int dictId = iBuf.get(i);
+        if (dictId != NULL_VALUE_INDEX) {
+          if (rawValue.equals(get(dictId))) {
+            return dictId;
+          }
+        }
+      }
+    }
+    if (_maxItemsInOverflowHash == 0) {
+      return NULL_VALUE_INDEX;
+    }
+    Integer dictId = valueToDictId.getOverflowMap().get(rawValue);
+    if (dictId == null) {
+      return NULL_VALUE_INDEX;
+    }
+    return dictId;
+  }
+
+  protected void indexValue(@Nonnull Object value) {
+    final long hashVal = Math.abs(value.hashCode());
+    ValueToDictId valueToDictId = _valueToDict.get();
+    final List<IntBuffer> iBufList = valueToDictId.getIBufList();
+
+    for (IntBuffer iBuf : iBufList) {
+      final int modulo = iBuf.capacity()/NUM_COLUMNS;
+      final int offsetInBuf = (int)(hashVal % modulo) * NUM_COLUMNS;
+      for (int i = offsetInBuf; i < offsetInBuf + NUM_COLUMNS; i++) {
+        final int dictId = iBuf.get(i);
+        if (dictId == NULL_VALUE_INDEX) {
+          setRawValueAt(_numEntries, value);
+          iBuf.put(i, _numEntries++);
+          return;
+        }
+        if (value.equals(get(dictId))) {
+          return;
+        }
+      }
+    }
+    // We know that we had a hash collision beyond the number of columns in the buffer.
+    Map<Object, Integer> overflowMap = valueToDictId.getOverflowMap();
+    if (_maxItemsInOverflowHash > 0) {
+      Integer dictId = overflowMap.get(value);
+      if (dictId != null) {
+        return;
+      }
+    }
+
+    setRawValueAt(_numEntries, value);
```

Comment:
we need to set the raw value *before* we update the reverse index, and hence it is here.